In [68]:
#Load the data in your code
import pandas as pd

data = pd.read_csv(r"C:\Users\Dennis\Desktop\Machine learning\HousePrices.csv")
data

price  lotsize  bedrooms  bathrooms  stories  driveway  recreation  \
0     42000     5850         3          1        2         1           0   
1     38500     4000         2          1        1         1           0   
2     49500     3060         3          1        1         1           0   
3     60500     6650         3          1        2         1           1   
4     61000     6360         2          1        1         1           0   
..      ...      ...       ...        ...      ...       ...         ...   
541   91500     4800         3          2        4         1           1   
542   94000     6000         3          2        4         1           0   
543  103000     6000         3          2        4         1           1   
544  105000     6000         3          2        2         1           1   
545  105000     6000         3          1        2         1           0   

     fullbase  gasheat  aircon  garage  prefer  
0           1        0       0       1       0  
1           0        0       0       0       0  
2           0        0       0       0       0  
3           0        0       0       0       0  
4           0        0       0       0       0  
..        ...      ...     ...     ...     ...  
541         0        0       1       0       0  
542         0        0       1       0       0  
543         0        0       1       1       0  
544         0        0       1       1       0  
545         0        0       1       1       0  

[546 rows x 12 columns]

Mean, median, max and mean values of columns whose their data type is integers

In [69]:
#Price column
#Mean
mean_price = data['price'].mean()
print("The mean of the price of houses is:", mean_price)

#Median
median_price = data['price'].median()
print("The median of the price of the houses is:",median_price)

# maximum
Max_price = data['price'].max()
print("The maximum price of the Houses is:",Max_price)

#Minimum
Min_price = data['price'].min()
print("The minimum price of the Houses is:",Min_price)

The mean of the price of houses is: 68121.59706959708
The median of the price of the houses is: 62000.0
The maximum price of the Houses is: 190000
The minimum price of the Houses is: 25000


In [70]:
#Lotsize column
#Mean
mean_lotsize = data['lotsize'].mean()
print("The mean of the lotsize of houses is:", mean_lotsize)
#Median
median_lotsize = data['lotsize'].median()
print("The median of the lotsize of the houses is:",median_lotsize)

# maximum
Max_lotsize = data['lotsize'].max()
print("The maximum lotsize of the Houses is:",Max_lotsize)

#Minimum
Min_lotsize = data['lotsize'].min()
print("The minimum lotsize of the Houses is:",Min_lotsize)

The mean of the lotsize of houses is: 5150.2655677655675
The median of the lotsize of the houses is: 4600.0
The maximum lotsize of the Houses is: 16200
The minimum lotsize of the Houses is: 1650


In [71]:
#bedrooms column
#Mean
mean_bedrooms = data['bedrooms'].mean()
print("The mean of the bedrooms of houses is:", mean_bedrooms)
#Median
median_bedrooms = data['bedrooms'].median()
print("The median of the bedrooms of the houses is:",median_bedrooms)

# maximum
Max_bedrooms = data['bedrooms'].max()
print("The maximum bedrooms of the Houses is:",Max_bedrooms)

#Minimum
Min_bedrooms = data['bedrooms'].min()
print("The minimum bedrooms of the Houses is:",Min_bedrooms)

The mean of the bedrooms of houses is: 2.9652014652014653
The median of the bedrooms of the houses is: 3.0
The maximum bedrooms of the Houses is: 6
The minimum bedrooms of the Houses is: 1


In [72]:
#Stories column
#Mean
mean_stories = data['stories'].mean()
print("The mean number of stories of the houses is:", mean_stories)
#Median
median_stories = data['stories'].median()
print("The median of the stories of the houses is:",median_stories)

# maximum
Max_stories = data['stories'].max()
print("The maximum stroies of the Houses is:",Max_stories)

#Minimum
min_stories = data['stories'].min()
print("The minimum stories of the Houses is:",min_stories)

The mean number of stories of the houses is: 1.8076923076923077
The median of the stories of the houses is: 2.0
The maximum stroies of the Houses is: 4
The minimum stories of the Houses is: 1


Mean, median , maximum and minimum values of the columns whose data types is strings

First i will change their values to integers in their original file by changing 'yes' to 1 and 'no' to 0
Now that i have changed the value of the columns with strings. Now i can find the mean, max and min. 

The above output means that many of the houses have recreation.

**percentage of houses in the data with Driveway, Gas-Heat and Air-conditioning present**

**DRIVEWAY**

In [37]:
#Houses with Driveways
#First get the number of houses with driveways
filtered_driveway = data.loc[data['driveway'] == 'yes', 'driveway']

#Convert the number to percent
percentage_filtered_driveway = len(filtered_driveway) / len(data['driveway']) * 100

print("The percentage of houses with driveways is:",percentage_filtered_driveway,"%")

The percentage of houses with driveways is: 85.8974358974359 %


**GAS_HEATS**

In [38]:
#Houses with gasheat
#First get the number of houses with gasheat
filtered_gasheat = data.loc[data['gasheat'] == 'yes', 'gasheat']

#Convert the number to percent
percentage_filtered_gasheat = len(filtered_gasheat) / len(data['gasheat']) * 100

print("The percentage of houses with gasheat is:",percentage_filtered_gasheat,"%")

The percentage of houses with gasheat is: 4.5787545787545785 %


**AIR_CONDITIONING**

In [39]:
#Houses with airconditioning
#First get the number of houses with airconditioning
filtered_aircon = data.loc[data['aircon'] == 'yes', 'aircon']

#Convert the number to percent
percentage_filtered_aircon = len(filtered_aircon) / len(data['aircon']) * 100

print("The percentage of houses with aircon is:",percentage_filtered_aircon,"%")

The percentage of houses with aircon is: 31.684981684981683 %


**RECREATION**

In [79]:
#Find the number of houses with recreations
filtered_recreation_yes = data.loc[data['recreation'] == 1, 'recreation']

#Convert the numbers into percentage
percentage_filtered_recreation_yes = len(filtered_recreation_yes) / len(data['recreation']) * 100
print("The percentage of of houses with recreation is",percentage_filtered_recreation_yes,"%")

#Find the number of houses with no recreations
filtered_recreation_no = data.loc[data['recreation'] == 0, 'recreation']

#Convert the number into percentage
percentage_filtered_recreation_no = len(filtered_recreation_no) / len(data['recreation']) * 100
print("The percentage of houses with  no recreations is :",percentage_filtered_recreation_no, "%")

The percentage of of houses with recreation is 17.765567765567766 %
The percentage of houses with  no recreations is : 82.23443223443223 %


**FULL-BASE**

In [80]:
#Find the number of houses with fullbase
filtered_fullbase_yes = data.loc[data['fullbase'] == 1, 'fullbase']

#Convert the numbers into percentage
percentage_filtered_fullbase_yes = len(filtered_fullbase_yes) / len(data['fullbase']) * 100
print("The percentage of houses with fullbase is",percentage_filtered_fullbase_yes,"%")

#Find the number of houses with no recreations
filtered_fullbase_no = data.loc[data['fullbase'] == 0, 'fullbase']

#Convert the number into percentage
percentage_filtered_fullbase_no = len(filtered_fullbase_no) / len(data['fullbase']) * 100
print("The percentage of houses with  no fullbase is :",percentage_filtered_fullbase_no, "%")

The percentage of houses with fullbase is 34.98168498168498 %
The percentage of houses with  no fullbase is : 65.01831501831502 %


**LINEAR REGRESSION MODEL THAT PREDICTS WHETHER THE PRICE OF THE HOUSE IS DETERMINED BY THE NUMBER OF BEDROOMS**

In [59]:
#Importing the neccesary libraries

from sklearn import linear_model
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

#independent variable
X = data[['bedrooms']]
y= data['price']

#Create a regresion model
model = linear_model.LinearRegression()

#train the regressin model
model.fit(X, y)

#print the coefficient and the intercept
print("co_efficient",model.coef_[0])
print("intercept:",model.intercept_)

#Make predictions in the model
y_pred = model.predict(X)

#Find the mean squared error an the square error
mse = mean_squared_error(y, y_pred)
r2 = r2_score(y, y_pred)

#Print the mean square error and the square error
print("R-Square: ",r2)
print("Mean Square Error: ",mse)

print(classification_report(y_test, y_pred))

co_efficient 13269.980086649637
intercept: 28773.432673469302
R-Square:  0.13428366667190073
Mean Square Error:  616153441.1715435


In [54]:
import statsmodels.api as sm

# Assuming you've already created and trained the model as shown in your code
# model = linear_model.LinearRegression()
# model.fit(X, y)

# Add a constant (intercept) to the independent variable
X = sm.add_constant(X)

# Create and fit an OLS (Ordinary Least Squares) model
model = sm.OLS(y, X).fit()

# Get the p-value for the coefficient of bedrooms
p_value = model.pvalues['bedrooms']

# Print the p-value
print("P-value for the coefficient:", p_value)


P-value for the coefficient: 8.533170549383786e-19


The coefficeint means that for every one bedroom addes to the house, the price increases by 13269 of the currency used, since it is a positive currency

The intercept means that the price of the house when the bedrooms are 0.

With a p-value of 8, there is a very little evidence to reject the null hypothensis. This only leaves us with the chance of alterntive hypothensis that states that there is a significant change in price of the houses as the number of bedrooms increases.

A low r-square value like this (0.1342836666719007) suggests that the number of bedrooms alone may not be string to predict the price pf houses. There are likely t be many other facts not considered in our model that influence house prices.

**multiple linear regression model by including all variables as predictors of house prices (response variable) and observe the effect on the house prices.**

In [67]:
# Import the required libraries
import pandas as pd
from sklearn import linear_model
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

# Change the values of the columns that have strings into integers to simplify learning of the model.
house = pd.get_dummies(data, columns=['driveway', 'recreation', 'fullbase', 'gasheat', 'aircon', 'prefer'])

# Independent variable
x = pd.concat([data[['lotsize', 'bedrooms', 'bathrooms', 'stories', 'garage']], house], axis=1)
y = data['price']

# Split the data into train and test (80% training and 20% testing)
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

# Create a regression model
model1 = linear_model.LinearRegression()

# Train the model
model1.fit(x_train, y_train)

# Test the model
y_pred1 = model1.predict(x_test)

#Find the mean square error and the square error

mse1 = mean_squared_error(y_test, y_pred1)
r2_1 = r2_score(y_test, y_pred1)

#Print them
print("R-Square: ",r2_1)
print("Mean Square Error: ",mse1)

R-Square:  1.0
Mean Square Error:  3.741864298239878e-22
